api/utils/accessio/chunkedreader.go

```diff
@@ -0,0 +1,118 @@
+package accessio
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"github.com/mandelsoft/goutils/general"
+)
+
+// ChunkedReader splits a reader into several
+// logical readers with a limited content size.
+// Once the reader reaches its limits it provides
+// a io.EOF.
+// It can be continued by Calling Next, which returns
+// whether a follow-up is required or not.
+type ChunkedReader struct {
+	lock   sync.Mutex
+	reader io.Reader
+	buffer *bytes.Buffer
+	size   int64
+	chunk  int
+	read   int64
+	err    error
+
+	preread uint
+}
+
+var _ io.Reader = (*ChunkedReader)(nil)
+
+func NewChunkedReader(r io.Reader, chunk int64, preread ...uint) *ChunkedReader {
+	return &ChunkedReader{
+		reader:  r,
+		size:    chunk,
+		preread: general.OptionalDefaulted(8096, preread...),
+	}
+}
+
+func (c *ChunkedReader) Read(p []byte) (n int, err error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.read >= c.size {
+		return 0, io.EOF
+	}
+	if c.read+int64(len(p)) > c.size {
+		p = p[:c.size-c.read] // read at most rest of chunk size
+	}
+	if c.buffer != nil && c.buffer.Len() > 0 {
+		// first, consume from buffer
+		n, _ := c.buffer.Read(p)
```

Comment:
Don't ignore the read error please.

Reply:
Hmm, the buffer `Read`only returns an EOF error, which is handled later.
But I'll replace the length check by an error check.